### PR TITLE
deal with IPv6 in DHT

### PR DESCRIPTION
**Pull request by NamsooCho**

### CODE REVIEW

File: src/toxcore/dht/dht_node.rs [hunk 1]
```diff
@@ -102,10 +122,26 @@ impl DhtNode {
 
     /// check it the node is timed out
     pub fn is_bad_node_timed_out(&self, server: &Server) -> bool {
-        self.last_resp_time.elapsed() > Duration::from_secs(server.config.bad_node_timeout)
+        self.last_resp_time_v4.elapsed() > Duration::from_secs(server.config.bad_node_timeout) &&
```
Comment (kpp) [inline src/toxcore/dht/dht_node.rs]: You may reuse fn `calc_status` here to reduce the code

File: src/toxcore/dht/server/mod.rs [hunk 2]
```diff
@@ -604,6 +609,25 @@ impl Server {
 
     /// actual send method
     fn send_to(&self, addr: SocketAddr, packet: DhtPacket) -> IoFuture<()> {
+        if !self.is_ipv6_mode { // DHT node is running in ipv4 mode
```
Comment (kpp) [inline src/toxcore/dht/server/mod.rs]: Please don't do `if !self.is_ipv6_mode`. Better code it like: `if self.is_ipv6_mode {  // DHT node is running in ipv6 mode }`

**Comment by kpp** [comment]
```
     Running `target/debug/tox-node --bootstrap-node 1D5A5F2F5D6233058BF0259B09622FB40B482E4FA0931EB8FD3AB8E7BF7DAF6F '198.98.51.198:33445' --bootstrap-node B71E91E2F5029B0A84D3B1136319CDD3D1DB6D3702B6CEFA66A4BEB25A635916 '[2a06:f901:1:100::98]:33445' --keys-file keys --udp-address '[::]:33445' --no-lan`
 INFO 2018-07-08T07:32:10Z: tox_node: DHT public key: AFBEB56D5ED57C8507ABC0EA4D70974D8806D4AE1E72FACB25C7C4F4572B5667
 INFO 2018-07-08T07:32:10Z: tox_node: Running DHT server on [::]:33445
 INFO 2018-07-08T07:32:10Z: tox_node: Running TCP relay on 0.0.0.0:33445
ERROR 2018-07-08T07:32:22Z: tox_node: Failed to handle packet: Custom { kind: Other, error: StringError("DHT node is running in ipv4 mode but target node\'s socket is ipv6 address") }
ERROR 2018-07-08T07:32:22Z: tox_node: Failed to handle packet: Custom { kind: Other, error: StringError("DHT node is running in ipv4 mode but target node\'s socket is ipv6 address") }
^C
```

**Comment by NamsooCho** [comment]
Default running mode is IPv4. If the user of DHT want to run it in IPv6 mode, he(she) should call enable_ipv6_mode().

https://github.com/tox-rs/tox/blob/97fced479833c9be6c6b9315594c00f53fbf6a34/src/toxcore/dht/server/mod.rs#L194

https://github.com/tox-rs/tox/blob/97fced479833c9be6c6b9315594c00f53fbf6a34/examples/dht_server.rs#L106-L109

File: src/toxcore/dht/server/mod.rs (hunk 2 above)
Comment (NamsooCho) [reply src/toxcore/dht/server/mod.rs]: done

**Comment by kpp** [comment]
```
diff --git a/Cargo.toml b/Cargo.toml
index e644cd7..45cfee6 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -23,4 +23,4 @@ itertools = "0.7"
 log = "0.4"
 tokio = "0.1.7"
 tokio-codec = "0.1"
-tox = { git = "https://github.com/tox-rs/tox" }
+tox = { git = "https://github.com/tox-rs/tox", branch="ipv6" }
diff --git a/src/main.rs b/src/main.rs
index 96dff43..9c6ca46 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -258,6 +258,12 @@ fn run_udp(cli_config: &CliConfig, dht_pk: PublicKey, dht_sk: SecretKey, udp_oni
     server.enable_lan_discovery(cli_config.lan_discovery_enabled);
     server.set_tcp_onion_sink(udp_onion.tx);
 
+    // set DHT server running mode to IPv6
+    if udp_addr.is_ipv6() {
+        server.enable_ipv6_mode(true);
+    }
+
+
     let server_c = server.clone();
     let udp_onion_future = udp_onion.rx
         .map_err(|()| Error::from(ErrorKind::UnexpectedEof))


humbug@home:~/tox-node$ cargo run -- --bootstrap-node 1D5A5F2F5D6233058BF0259B09622FB40B482E4FA0931EB8FD3AB8E7BF7DAF6F 198.98.51.198:33445 --bootstrap-node B71E91E2F5029B0A84D3B1136319CDD3D1DB6D3702B6CEFA66A4BEB25A635916 [2a06:f901:1:100::98]:33445 --keys-file keys --udp-address [::]:33445 --no-lan 
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
     Running `target/debug/tox-node --bootstrap-node 1D5A5F2F5D6233058BF0259B09622FB40B482E4FA0931EB8FD3AB8E7BF7DAF6F '198.98.51.198:33445' --bootstrap-node B71E91E2F5029B0A84D3B1136319CDD3D1DB6D3702B6CEFA66A4BEB25A635916 '[2a06:f901:1:100::98]:33445' --keys-file keys --udp-address '[::]:33445' --no-lan`
 INFO 2018-07-08T21:32:01Z: tox_node: DHT public key: AFBEB56D5ED57C8507ABC0EA4D70974D8806D4AE1E72FACB25C7C4F4572B5667
 INFO 2018-07-08T21:32:01Z: tox_node: Running DHT server on [::]:33445
 INFO 2018-07-08T21:32:01Z: tox_node: Running TCP relay on 0.0.0.0:33445
thread 'main' panicked at 'Execution was terminated with error: Custom { kind: Other, error: Compat { error: Os { code: 101, kind: Other, message: "Network is unreachable" } } }', libcore/result.rs:945:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

```

**Comment by kurnevsky** [comment]
What about `KBucket::try_add`? Should it merge `IPv4` and `IPv6` nodes in stead of replacing?

**Comment by NamsooCho** [comment]
Ah. That's right.

File: src/toxcore/onion/onion_announce.rs [hunk 3]
```diff
@@ -103,7 +103,7 @@ Length   | Content
 `8`      | Unix time in seconds divided by PING_ID_TIMEOUT
 `32`     | `PublicKey` of sender
 `1`      | IP type of sender
-`8`      | `IpAddr` of sender
+`16`      | `IpAddr` of sender
```
Comment (kpp) [inline src/toxcore/onion/onion_announce.rs]: Fix indent

File: src/toxcore/dht/server/mod.rs [hunk 4]
```diff
@@ -506,8 +514,10 @@ impl Server {
     // actual sending function of NatPingRequest.
     fn send_nat_ping_req_inner(&self, friend: &DhtFriend, nat_ping_req_packet: DhtPacket) -> IoFuture<()> {
         let nats_sender = friend.close_nodes.nodes.iter()
-            .map(|node| {
-                self.send_to(node.saddr, nat_ping_req_packet.clone())
+            .map(|node| node.get_socket_addr(self.is_ipv6_mode))
+            .filter(|addr| addr.is_some())
+            .map(|addr| {
```
Comment (kurnevsky) [inline src/toxcore/dht/server/mod.rs]: filter_map

File: src/toxcore/dht/server/mod.rs [hunk 5]
```diff
@@ -691,10 +720,30 @@ impl Server {
         };
 
         let close_nodes = close_nodes.get_closest(&self.pk, IsGlobal::is_global(&addr.ip()));
+
+        let mut collected_bucket = Bucket::new(Some(4));
+
+        close_nodes.iter()
+            .for_each(|node| {
+                collected_bucket.try_add(&self.pk, node);
```
Comment (kurnevsky) [inline src/toxcore/dht/server/mod.rs]: Why `self.pk` is used? Shouldn't it be payload.pk? And why do we need to iterate friends here?

File: src/toxcore/onion/onion_announce.rs [hunk 6]
```diff
@@ -308,7 +307,7 @@ impl OnionAnnounce {
         let response_payload = OnionAnnounceResponsePayload {
             announce_status,
             ping_id_or_pk,
-            nodes: kbucket.get_closest(&payload.search_pk, IsGlobal::is_global(&addr.ip()))
+            nodes: kbucket.get_closest(&payload.search_pk, false)
```
Comment (kurnevsky) [inline src/toxcore/onion/onion_announce.rs]: Shouldn't it be `true`?

File: src/toxcore/dht/server/mod.rs (hunk 5 above)
Comment (NamsooCho) [reply src/toxcore/dht/server/mod.rs]: payload.pk should be used.
When we respond with NodesResponse, we should search not only close_list but also each friend's close_list. It was omitted in previous PR.

File: src/toxcore/dht/dht_friend.rs [hunk 7]
```diff
@@ -412,6 +415,6 @@ mod tests {
             saddr: "127.0.0.1:33445".parse().unwrap(),
         }));
 
-        assert_eq!(friend.get_addrs_of_clients(), vec!["127.0.0.1:33445".parse().unwrap()]);
+        assert_eq!(friend.get_addrs_of_clients(true), vec!["127.0.0.1:33445".parse().unwrap()]);
```
Comment (kpp) [inline src/toxcore/dht/dht_friend.rs]: Add more test cases

File: src/toxcore/dht/dht_node.rs [hunk 8]
```diff
@@ -102,10 +129,33 @@ impl DhtNode {
...
     pub fn is_bad_node_timed_out(&self, server: &Server) -> bool {
-        self.last_resp_time.elapsed() > Duration::from_secs(server.config.bad_node_timeout)
+        self.calc_status(Duration::from_secs(server.config.bad_node_timeout)) == NodeStatus::Bad
     }
 
-
+    /// return SocketAddr for DhtNode
+    pub fn get_socket_addr(&self, is_ipv6_mode: bool) -> Option<SocketAddr> {
+        if is_ipv6_mode {
+            if self.saddr_v6.is_some() && self.saddr_v4.is_some() {
```
Comment (kurnevsky) [inline src/toxcore/dht/dht_node.rs]: Pattern matching here can help to avoid all these `unwrap`s.

**Pull request merged**